polytope_server/worker/worker.py

```diff
@@ -211,81 +215,105 @@ def run(self):
+                                self.staging.delete(id)
+
+                    # upload result data
+                    if datasource is not None:
+                        # Span for uploading data
+                        with create_new_span_internal("Uploading result data", request_id=id):
+                            request.url = self.staging.create(id, datasource.result(request), datasource.mime_type())
+                            # Getting key (name + ext) from url
+                            object_id = id + ("." + request.url.split("/")[-1].split(".")[-1])
+                            # Getting data size in bytes
+                            _, size = self.staging.stat(object_id)
+                            process_span.set_attributes({
+                                "polytope.request.url": request.url,
+                                "polytope.request.size": size
+                            })
+
+            except Exception as e:
+                request.user_message += f"Failed to finalize request: [{str(type(e))}] {str(e)}"
+                logging.info(request.user_message, extra={"request_id": id})
+                logging.exception("Failed to finalize request", extra={"request_id": id, "exception": str(e)})
+                set_span_error(process_span, e)
+                raise
+
+            # Guarantee destruction of the datasource
+            finally:
+                if datasource is not None:
+                    datasource.destroy(request)
+
+            if datasource is None:
+                request.user_message += "Failed to process request."
+                logging.info(request.user_message, extra={"request_id": id})
+                set_span_error(process_span, Exception(request.user_message))
+                raise Exception("Failed to process request.")
+            else:
+                request.user_message += "Success"
+
+            return
 
     def fetch_input_data(self, url):
         """Downloads input data from external URL or staging"""
-        if url != "":
-            try:
-                response = requests.get(url, proxies=self.proxies)
-                response.raise_for_status()
-            except (
-                requests.exceptions.ConnectionError,
-                requests.exceptions.HTTPError,
-            ):
-                logging.info("Retrying requests.get without proxies after failure")
-                response = requests.get(url)
-                response.raise_for_status()
-
-            if response.status_code == 200:
-                logging.info("Downloaded data of size {} from {}".format(sys.getsizeof(response._content), url))
-                return response._content
-            else:
-                raise Exception(
-                    "Could not download data from {}, got {} : {}".format(url, response.status_code, response._content)
-                )
-        return None
+        with create_new_span_internal("Fetching input data") as fetch_input_span:
```

Comment:
Also you do not set request_id here

Reply:
Same as before parent span has the request_id

Reply:
Do you think we should set the request_id attribute in all spans? this is an interesting point, the idea of setting the request id into the span as attribute is for later be able to filter easy from the aws dashboards, however, now the request_id appears in the main span of each service and not in their child

Reply:
If otel will group the children to the parents, then we probably don't need the request_id to be plumbed into all of these. Just making sure that we don't end up losing part of the trace.